core/src/main/scala/chisel3/package.scala

```diff
@@ -210,29 +211,116 @@ package object chisel3 {
+      // Only thing present is literal % if any which should ideally be with %%. 
+      // If not - then flag an error. 
+      // Return seq of Printables (either PString or Percent or both - nothing else
+      def percentSplitter(s : String): Seq[Printable] = {
+          if(s.isEmpty) return Seq(PString(""))
+	        val pieces = s.split("%%").toList.flatMap { p =>
+	          if (p.contains('%')) throw new UnknownFormatConversionException("Un-escaped % found")
+	          // Wrap in PString and intersperse the escaped percentages
+	          Seq(Percent, PString(p))
+	        }
+	        if(pieces.isEmpty) Seq(Percent) else pieces.tail  // Don't forget to drop the extra percent we put at the beginning
+        }
+
+      def extractFormatSpecifier(part: String): (Option[String], String) = {
+         // Check if part starts with a format specifier (with % - disambiguate with literal % checking the next character if needed to be %)
+          // In the case of %f specifier there is a chance that we need more information - so capture till the 1st letter (a-zA-Z).
+          // Example cf"This is $val%2.2f here" - parts - Seq("This is ","%2.2f here") - the format specifier here is %2.2f.
+          val idx_of_fmt_str =
+            if (
+              !part.isEmpty() && part.charAt(0) == '%' && (part.size == 1 || (part.size >= 2 && part.charAt(1) != '%'))
+            ) part.indexWhere { _.isLetter }
+            else -1
+
+          val fmt = if (idx_of_fmt_str >= 0) Some(part.substring(0, idx_of_fmt_str + 1)) else None
+
+          val formatRemovedString = part.zipWithIndex.collect { case (p, idx) if idx > idx_of_fmt_str => p }.mkString
+
+          (fmt,formatRemovedString)
+ 
       }
+
+      sc.checkLengths(args) // Enforce sc.parts.size == pargs.size + 1
       val parts = sc.parts.map(StringContext.treatEscapes)
-      // Zip sc.parts and pargs together ito flat Seq
-      // eg. Seq(sc.parts(0), pargs(0), sc.parts(1), pargs(1), ...)
-      val seq = for { // append None because sc.parts.size == pargs.size + 1
-        (literal, arg) <- parts.zip(pargs :+ None)
-        optPable <- Seq(Some(PString(literal)), arg)
-        pable <- optPable // Remove Option[_]
-      } yield pable
-      Printables(seq)
+      // The 1st part is assumed never to contain a format specifier.
+      // If the 1st part of a string is an argument - then the 1st part will be an empty String.
+      // So we need to parse parts following the 1st one to get the format specifiers if any
+      val partsAfterFirst = parts.tail
+
+      // Align parts to their potential specifiers
+      val pables = partsAfterFirst.zip(args).flatMap {case (part, arg) => {
+          val (fmt,modP) = extractFormatSpecifier(part)
+          val fmtArg: Printable = arg match {
+            case d: Data => {
+              fmt match {
+                case Some("%n") => Name(d)
+                case Some("%N") => FullName(d)
+                case Some("%s") => PString(d.toString)
```

Comment:
```suggestion
```
Upon further reflection, we should reserve `%s` for String's that map down to Verilog. Let's not add this right now.